exec/java-exec/src/main/java/org/apache/drill/exec/rpc/security/plain/PlainServer.java

```diff
@@ -0,0 +1,174 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.rpc.security.plain;
+
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.auth.callback.NameCallback;
+import javax.security.auth.callback.PasswordCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.sasl.AuthorizeCallback;
+import javax.security.sasl.Sasl;
+import javax.security.sasl.SaslException;
+import javax.security.sasl.SaslServer;
+import javax.security.sasl.SaslServerFactory;
+import java.io.IOException;
+import java.security.Provider;
+import java.util.Map;
+
+/**
+ * Plain SaslServer implementation. See https://tools.ietf.org/html/rfc4616
+ */
+public class PlainServer implements SaslServer {
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(PlainServer.class);
+
+  public static class PlainServerFactory implements SaslServerFactory {
+
+    @Override
+    public SaslServer createSaslServer(final String mechanism, final String protocol, final String serverName,
+                                       final Map<String, ?> props, final CallbackHandler cbh)
+        throws SaslException {
+      return "PLAIN".equals(mechanism) ?
+          props == null || "false".equals(props.get(Sasl.POLICY_NOPLAINTEXT)) ?
+              new PlainServer(cbh) :
+              null :
+          null;
+    }
+
+    @Override
+    public String[] getMechanismNames(final Map<String, ?> props) {
+      return props == null || "false".equals(props.get(Sasl.POLICY_NOPLAINTEXT)) ?
+          new String[]{"PLAIN"} :
+          new String[0];
+    }
+  }
+
+  @SuppressWarnings("serial")
+  public static class PlainServerProvider extends Provider {
+
+    public PlainServerProvider() {
+      super("PlainServer", 1.0, "PLAIN SASL Server Provider");
+      put("SaslServerFactory.PLAIN", PlainServerFactory.class.getName());
+    }
+  }
+
+  private CallbackHandler cbh;
+  private boolean completed;
+  private String authorizationID;
+
+  PlainServer(final CallbackHandler cbh) throws SaslException {
+    if (cbh == null) {
+      throw new SaslException("PLAIN: A callback handler must be specified.");
+    }
+    this.cbh = cbh;
+  }
+
+  @Override
+  public String getMechanismName() {
+    return "PLAIN";
+  }
+
+  @Override
+  public byte[] evaluateResponse(byte[] response) throws SaslException {
+    if (completed) {
+      throw new IllegalStateException("PLAIN authentication already completed");
+    }
+
+    if (response == null) {
+      throw new SaslException("Received null response");
+    }
+
+    completed = true;
```

Comment:
'completed' should be set at the end of the whole authentication since it is the critical flag to decide if the user was authenticated successfully.